crates/sp-subspace-mmr/src/host_functions.rs

```diff
@@ -56,4 +63,16 @@ where
             extrinsics_root: H256::from_slice(header.extrinsics_root().as_ref()),
         })
     }
+
+    fn verify_mmr_proof(&self, leaves: Vec<EncodableOpaqueLeaf>, encoded_proof: Vec<u8>) -> bool {
```

Comment:
But why do you need to verify the proof using consensus chain on domain? The proof structure is known, the proof verification algorithm is known as well, why not embed verification logic into domain runtime just like it is done on consensus side? It is just a function call.

Reply:
> But why do you need to verify the proof using consensus chain on domain? The proof structure is known, the proof verification algorithm is known as well, why not embed verification logic into domain runtime just like it is done on consensus side? It is just a function call.

While the proof structure is known, this is not a stateless call. `Mmr::verify_proof` essentially uses the offchain storage to get the peaks of MMR if not available on runtime to derive the mmr root. If we re-use the function here, we would still require host function call that can get those peaks into domain runtime from consensus chain storage.

There is `Mmr::verify_leaves_stateless` where proof itself carry the root. But in order to verify the root, we would need the host function to check if the consensus hash which was used to generate proof is canonical and then fetch the mmr root to cross check. I was planning to use this approach after looking at this pr #2502 since the state is always available but IIUC we would still require block data which is not gauranteed to be avaible.

So, I have decided to use the current approach since our reference implementation is must have the necessary peaks that might be missing from the consnesus runtime all the time and there is no need to worry about the state or block availability here. 
Either case, we would still require a Host function for domains to verify such mmr proof